app/src/main/kotlin/com/simplemobiletools/flashlight/activities/MainActivity.kt

```diff
@@ -212,121 +252,78 @@ class MainActivity : SimpleActivity() {
-        binding.brightnessBar.apply {
-            max = mCameraImpl?.getMaximumBrightnessLevel() ?: MIN_BRIGHTNESS_LEVEL
-            progress = mCameraImpl?.getCurrentBrightnessLevel() ?: MIN_BRIGHTNESS_LEVEL
-            onSeekBarChangeListener { level ->
-                val newLevel = level.coerceAtLeast(MIN_BRIGHTNESS_LEVEL)
-                mCameraImpl?.updateBrightnessLevel(newLevel)
-                config.brightnessLevel = newLevel
+            val permission = Manifest.permission.CAMERA
+            if (ContextCompat.checkSelfPermission(this, permission) == PackageManager.PERMISSION_GRANTED) {
+                cameraPermissionGranted(isSOS)
+            } else {
+                launcher.launch(permission)
             }
         }
     }
 
     private fun cameraPermissionGranted(isSOS: Boolean) {
         if (isSOS) {
-            val isSOSRunning = mCameraImpl!!.toggleSOS()
-            binding.sosBtn.setTextColor(if (isSOSRunning) getProperPrimaryColor() else getContrastColor())
-        } else if (mCameraImpl!!.toggleStroboscope()) {
-            binding.apply {
-                stroboscopeBar.beInvisibleIf(stroboscopeBar.isVisible())
-                changeIconColor(if (stroboscopeBar.isVisible()) getProperPrimaryColor() else getContrastColor(), stroboscopeBtn)
-            }
+            viewModel.enableSos()
+        } else {
+            viewModel.enableStroboscope()
         }
     }
 
-    private fun getContrastColor() = getProperBackgroundColor().getContrastColor()
-
-    private fun releaseCamera() {
-        mCameraImpl?.releaseCamera()
-        mCameraImpl = null
-    }
-
-    private fun showSleepTimer(force: Boolean = false) {
-        val alarmManager = getSystemService(Context.ALARM_SERVICE) as AlarmManager
-        if (isSPlus() && !alarmManager.canScheduleExactAlarms() && !force) {
-            PermissionRequiredDialog(
-                this,
-                com.simplemobiletools.commons.R.string.allow_alarm_sleep_timer,
-                positiveActionCallback = { openRequestExactAlarmSettings(baseConfig.appId) },
-                negativeActionCallback = { showSleepTimer(true) }
-            )
-            return
-        }
-
+    @Composable
+    private fun SleepTimerRadioDialog(
+        alertDialogState: AlertDialogState,
+        customAlertDialogState: AlertDialogState
+    ) {
         val items = ArrayList(listOf(10, 30, 60, 5 * 60, 10 * 60, 30 * 60).map {
             RadioItem(it, secondsToString(it))
         })
 
-        if (items.none { it.id == config.lastSleepTimerSeconds }) {
-            items.add(RadioItem(config.lastSleepTimerSeconds, secondsToString(config.lastSleepTimerSeconds)))
+        if (items.none { it.id == preferences.lastSleepTimerSeconds }) {
+            items.add(RadioItem(preferences.lastSleepTimerSeconds, secondsToString(config.lastSleepTimerSeconds)))
         }
 
         items.sortBy { it.id }
         items.add(RadioItem(-1, getString(R.string.custom)))
```

Comment:
`remember` thes

Reply:
I think this should be alright for this, right?
https://github.com/SimpleMobileTools/Simple-Flashlight/pull/213/commits/af6691c06a653b1ff30f28f79ab436147856534f

Reply:
There's no need of an arrayList, you can use `remember { mutableStateListOf<RadioItem>() }`